substratevm/src/com.oracle.svm.graal/src/com/oracle/svm/graal/isolated/IsolatedSpeculationLog.java

```diff
@@ -48,19 +53,89 @@ public void collectFailedSpeculations() {
         collectFailedSpeculations0(IsolatedCompileContext.get().getClient(), handle);
     }
 
+    private static byte[] encodeAsByteArray(SpeculationReason reason) {
+        int groupId = 0;
+        Object[] context = null;
+        try {
+            final Field groupIdField;
```

Comment:
If I understand the review conversation, these reflective accesses will be converted to non-reflective accesses during image building. A comment should be added somewhere here explaining that. Does the added unit test check that this happens?

Reply:
> Does the added unit test check that this happens?

Not at its current state. It cannot be compiled with native-image due to the invocation handler that is required to create the supplier that we pass to labsJDK's internal `encode`.
Would adding another test that only uses graalvm's implementations and running it with `native-unittest` be enough?

Reply:
If that tests the condition stated in the (to be added) comment, then yes.

Reply:
According to https://www.graalvm.org/reference-manual/native-image/Reflection/ 

>  If the target elements can be resolved, the calls are removed and instead the target elements are embedded in the code. If the target elements cannot be resolved, e.g., a class is not on the classpath or it does not declare a field/method/constructor, then the calls are replaced with a snippet that throws the appropriate exception at run time.

So to my understanding just compiling and running an image that covers the corresponding code segments should be enough.

Is there any better way to achieve this?

Reply:
[Comment](https://github.com/oracle/graal/pull/2945/files#diff-603f78c0054580a456da1e1f763e13bd03cd60958a2f5358fb25509c268d57f6R62-R67) added [test](https://github.com/oracle/graal/pull/2945/files#diff-69e104d7b1d63ef862b22d172b7e6727584403f02e35ff8250084fc6dae8f51f) added.

Reply:
Indeed, if the members cannot be resolved directly and are not registered by other means, usually a configuration file, using them will fail.